pkg/container/versioned/value.go

```diff
@@ -0,0 +1,440 @@
+}
+
+func newVersionHold(version version, coordinator *Coordinator) *VersionHold {
+	// hold on maxVersion never expires
+	if version == maxVersion {
+		coordinator = nil
+	}
+	h := &VersionHold{version: version}
+	h.coordinator.Store(coordinator)
+	if coordinator != nil {
+		// Set a finalizer to catch unclosed holds. The finalizer function complains
+		// loudly, so that we do not rely the finalizer for closing.
+		runtime.SetFinalizer(h, versionHoldFinalizer)
+	}
+	return h
+}
+
+// Latest returns a VersionHold for the latest version of current/non-removed values
+// Only to be used in cases where the latest values are needed and no transactionality is required.
+func Latest() *VersionHold {
+	return newVersionHold(maxVersion, nil)
+}
+
+func (h *VersionHold) IsValid() bool {
+	return h != nil && (h.version == maxVersion || h.coordinator.Load() != nil)
+}
+
+// Close releases the held version for removal once no hold for it are no longer held.
+// This may not be called while holding any locks that the 'closer' function passed to the
+// coordinator may take!
+func (h *VersionHold) Close() (err error) {
+	if !h.IsValid() {
+		return errInvalidVersion
+	}
+	coordinator := h.coordinator.Load()
+	if coordinator != nil {
+		runtime.SetFinalizer(h, nil)
+		err = coordinator.releaseVersion(h.version)
+		h.coordinator.Store(nil)
+	}
+	return err
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) load() version {
+	return version(a.version.Load())
+}
+
+func (a *atomicVersion) store(version version) {
+	a.version.Store(uint64(version))
+}
+
+func (a *atomicVersion) compareAndSwap(old, new version) bool {
+	return a.version.CompareAndSwap(uint64(old), uint64(new))
+}
+
+type versionCount struct {
+	version version
+	count   int
+}
+
+// Coordinator defines a common version number space for multiple versioned.Values,
+// and provides facilities for cleaning out stale versions.
+// The Value's are not directly managed by the Coordinator, but all the values under coordination
+// should be cleaned by the 'cleaner' function given to the Coordinator.
+type Coordinator struct {
+	// version is the version number of the last applied change
+	version atomicVersion
+
+	// cleaner is called with the earliest version that must be kept
+	cleaner func(KeepVersion)
+
+	// mutex protects the rest of the fields
+	mutex lock.RWMutex
+
+	// versions is an ordered list of outstanding VersionHolds with a reference count.
+	// Outdated values can be cleaned when there are no outstanding VersionHolds for them.
+	versions []versionCount
+
+	// latest cleaned version
+	cleanedVersion version
+}
+
+// NewCoordinator returns a new Coordinator
+func NewCoordinator(cleaner func(KeepVersion)) *Coordinator {
+	return &Coordinator{
+		cleaner: cleaner,
+	}
+}
+
+// GetNextVersion returns NextVersion to be used when adding or removing values.
+//
+// Callers need to coordinate so that a single goroutine is performing modifications at any one
```

Comment:
Looking at this, I can see the transactionality clearly. Should the transaction be made explicit in the API? Explicit transactions in Go could look like in SQL https://pkg.go.dev/database/sql#Tx or in statedb: https://pkg.go.dev/github.com/cilium/statedb#ReadTxn

Reply:
Updated, now it looks like this:
```
// - tx := coordinator.PrepareNextVersion()
//   - value.SetAt(... , tx)
//   - value.RemoveAt(..., tx)
//   - ...
//
// - tx.Commit()
```